### PR TITLE
[bitnet] Support use_sub_norms=False (weight-quant-only BitLinear checkpoints)

### LINKED ISSUE #47957: [bitnet] Weight-quant-only BitLinear checkpoints load with sub-norms forced on
### Problem

`BitNetMLP.forward` always applies `self.ffn_sub_norm(...)` and `BitNetAttention.forward` always applies `self.attn_sub_norm(...)`, and both modules are unconditionally constructed in `__init__`. This is only correct for **full BitNet** checkpoints.

BitNet checkpoints actually come in two forward variants:

1. **Full BitNet** (`microsoft/bitnet-b1.58-2B-4T` shape): per-projection RMSNorm (`attn_sub_norm` / `ffn_sub_norm`) applied before the projection bias and the residual.
2. **Weight-quant-only**: trained through a BitLinear that skips the per-projection RMSNorm and activation quant entirely; weights are already ternary `{-1, 0, +1}` and the forward is a plain `nn.Linear`. The sub-norm modules are absent — and they cannot be "neutralised" by loading `weight=ones`, because an RMSNorm initialised to ones still normalises.

Loading a weight-quant-only checkpoint (e.g. `PeetPedro/quantal-ternary`, a ternary export of a continued-trained Qwen2.5-0.5B) therefore runs the sub-norms, producing logits that differ from the trained/deployed forward by ~10 in max-abs and flip the argmax (verified against a golden reference the Rust runner reproduces to 1e-5).

### Proposed fix

- `PR #47955`: adds `use_sub_norms: bool = True` to `BitNetConfig`; when `False`, the sub-norm modules are replaced by `nn.Identity` (built by a shared `_make_sub_norm` helper), so the normalisation is removed from the forward while the module graph and state dict stay uniform (Identity has no parameters). Default stays `True`, so existing model cards are unchanged.

### Context

- Coordination model: https://huggingface.co/PeetPedro/quantal-ternary (the checkpoint this enables)
- PR: https://github.com/huggingface/transformers/pull/47955

### PR DESCRIPTION
**Pull request by peterlodri-sec**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47955)

## Summary

Adds the `use_sub_norms` flag to `BitNetConfig` and honours it in
`BitNetMLP`/`BitNetAttention`. When `False`, the per-projection sub-layer
RMSNorm modules are replaced by `nn.Identity` (a shared `_make_sub_norm`
helper builds either), which is what weight-quant-only BitLinear
checkpoints require.

## Why

BitNet checkpoints come in two forward variants:

1. **Full BitNet** (`use_sub_norms=True`, default): per-projection RMSNorm
   (attn_sub_norm / ffn_sub_norm) applied before the projection bias and the
   residual — the `microsoft/bitnet-b1.58-2B-4T` shape.
2. **Weight-quant-only** (`use_sub_norms=False`): trained through a BitLinear
   that skips the per-projection RMSNorm and activation quant entirely; the
   weights are already ternary `{-1, 0, +1}` and the forward is a plain
   `nn.Linear`. The sub-norm must not normalise: an RMSNorm initialised to
   ones still normalises, so it cannot be neutralised by `weight=ones`.
   `nn.Identity` removes the normalisation from the forward while keeping the
   module graph and the state dict uniform (it has no parameters, so no
   `*_sub_norm.*` weights appear either way).

Today `BitNetMLP.forward` always runs `self.ffn_sub_norm(...)` and
`BitNetAttention.forward` always runs `self.attn_sub_norm(...)` on modules
unconditionally constructed in `__init__`. A weight-quant-only checkpoint
therefore loads with the sub-norms applied, producing logits that differ
from the trained/deployed forward by ~10 in max-abs and flip the argmax.

## Changes

- `configuration_bitnet.py`: add `use_sub_norms: bool = True` (docstring
  notes that loading a checkpoint *with* sub-norm weights into a
  `use_sub_norms=False` model leaves them unused and reported as such).
- `modular_bitnet.py`: `_make_sub_norm` helper + `nn.Identity` fallback; the
  forward paths stay branch-free.
- `modeling_bitnet.py`: generated mirror of the modular change.
- `tests/models/bitnet/test_modeling_bitnet.py`: new
  `test_model_without_sub_norms` (Identity modules, no `*_sub_norm.*` state
  dict keys, forward shapes) and `test_config_use_sub_norms_roundtrip`.

Backward compatible: the default stays `True`, so existing `bitnet` model
cards (which don't set `use_sub_norms`) are unchanged.

## Verification

A continued-trained Qwen2.5-0.5B weight-quant-only BitNet (ternary export,
`PeetPedro/quantal-ternary`) loaded into `BitNetForCausalLM` with
`use_sub_norms=False`, compared against a golden reference (the deployed
forward the Rust runner reproduces to 1e-5):

```
prompt 1: max_abs 6.497e-06  argmax 71703 = 71703  PASS
prompt 2: max_abs 6.676e-06  argmax 71703 = 71703  PASS
```

Before the fix (sub-norms forced on): max_abs ~10-20, argmax off.

Local: `pytest tests/models/bitnet/test_modeling_bitnet.py` — 146 passed,
132 skipped; ruff and modular-conversion checks pass.

## AI disclosure

This PR was drafted with assistance from an AI coding agent (opencode), and
further refined with an AI assistant (Crush). The human submitter reviewed
all changed lines and ran the verification above.
Coordination: https://huggingface.co/PeetPedro/quantal-ternary (the model
this enables). No overlapping open PR found for `use_sub_norms`.

Fixes #47957

### CODE REVIEW

**Comment by peterlodri-sec** [comment]
**Context — the HF articles this change is documented in**

This PR makes `PeetPedro/quantal-ternary` (a weight-quant-only BitNet b1.58
checkpoint) load correctly in transformers. The full engineering story lives
in the constellation's quantal thread; each article carries one question that
this change answers:

1. **[The third state is real](https://pocoo.vaked.dev/posts/2026-08-13-quantal-ternary-third-state-is-real)**
   — *"Does this model ever want a zero? If the trained weights are all at
   plus or minus one scale, b1.58 is doing b1 work, and the third state is a
   slot you are paying for in every file."* The checkpoint is weight-quant-only
   ternary with ~30% real zeros; the loader must not re-apply the sub-layer
   RMSNorm modules the deployed forward never had.

2. **[From 11.34 to 0.5597](https://pocoo.vaked.dev/posts/2026-08-12-quantal-ternary-11_34-to-0_63)**
   — *"What you evaluate is what you ship."* The trained forward *is* the
   deployed forward (weight-quant-only BitLinear). `use_sub_norms=False` is
   what makes the PyTorch loader agree with the deployed forward to
   **6.5e-06** instead of flipping the argmax at ~10 max-abs.

3. **[From 3.29 to 1.64](https://pocoo.vaked.dev/posts/2026-08-11-quantal-ternary-3_29-to-1_64)**
   — *"Why ternary, again?"* Because weight-quant-only keeps training ≡
   deployment; full BitLinear's per-projection RMSNorm is a forward that does
   not exist in the Rust runtime, so it must be absent (not neutralised) in
   the loader too.

Model card: https://huggingface.co/PeetPedro/quantal-ternary

**Comment by Rocketknight1** [comment]
No code agent PRs from first-time contributors please!
